src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs

```diff
@@ -48,5 +53,30 @@ MinDataRate IHttpMinRequestBodyDataRateFeature.MinDataRate
                 MinRequestBodyDataRate = value;
             }
         }
+
+        async Task IHttpResponseCompletionFeature.CompleteAsync()
+        {
```

Comment:
What would happen if this method was just verifying the content-length and call ProduceEnd()? It should initialize the response for you if the headers haven't been committed yet and will return the task from WriteSuffix(). This might help get rid of an unnecessary asyncstatemachine.